lib/adapter.js

```diff
@@ -425,44 +425,36 @@ AbstractPouchDB.prototype.compact =
   });
 });
 AbstractPouchDB.prototype._compact = function (opts, callback) {
-  var done = false;
-  var started = 0;
-  var copts = {
-    returnDocs: false
-  };
   var self = this;
-  var lastSeq;
-  function finish() {
-    upsert(self, '_local/compaction', function (doc) {
-      if (!doc.last_seq || doc.last_seq < lastSeq) {
-        doc.last_seq = lastSeq;
-        return doc;
-      }
-      return false; // somebody else got here first, don't update
-    }, function () {
-      callback(null, {ok: true});
+  var changesOpts = {
+    returnDocs: false,
+    last_seq: opts.last_seq || 0
+  };
+  var promise = utils.Promise.resolve();
+
+  function onChange(row) {
+    promise = promise.then(function () {
+      return self.compactDocument(row.id, 0);
     });
   }
-  if (opts.last_seq) {
-    copts.since = opts.last_seq;
-  }
-  function afterCompact() {
-    started--;
-    if (!started && done) {
-      finish();
-    }
-  }
-  function onChange(row) {
-    started++;
-    self.compactDocument(row.id, 0).then(afterCompact, callback);
```

Comment:
^ this was the source of the bug. We were calling `callback` for every `'change'` event, plus during the `'complete'` event.

This code was essentially just trying to do the compactions in order and then sequentially run `finish` at the end. So why not use a promise queue for that?